### PR TITLE
feat(@desktop): single application instance

### PR DESCRIPTION
**Pull request by pinebit**
As a user, I should not be able to run the second Status application instance, either intentionally or not, **for the same data directory** (the directory having your status account data).
The default data directory can be overridden with -d switch. 

If the user is attempted to start the second Status instance, the new process should be immediately terminated and the other (already running) instance should be activated (the main window should be made visible and brought to the foreground).

TODO: support passing command line arguments such as --url for deep links.

### CODE REVIEW

**Comment by borismelnik** [comment]
This window was made for custom toolbar and can be used in other points. The standard for single application instance is a single QApplication instance not window.

PS: Which task you try to implement here? maybe i don't understand.

**Comment by anastasiyaig** [comment]
@pinebit would u mind to put a description for the functionality that is planned to be introduced in this PR? Having empty descriptions is confusing and not informative. Thanks

**Comment by iurimatias** [comment]
later one we'll need to be able to do https://github.com/status-im/status-desktop/issues/2389

**Comment by richard-ramos** [comment]
Wouldnt this stop us from being able to login into more than one account?

**Comment by pinebit** [comment]
> This window was made for custom toolbar and can be used in other points. The standard for single application instance is a single QApplication instance not window.
> for example https://github.com/itay-grudev/SingleApplication
> 
> PS: Which task you try to implement here? maybe i don't understand.

`StatusWindow` is currently used in the single place in our code (main.qml), hence I decided not to over-engineering this and embed this functionality to `StatusWindow`. However, if you expect to use this component in other places in near future, then yes it is better to introduce custom `StatusApplication` that will implement this behavior.
This feature was implemented in addition to https://github.com/status-im/status-desktop/issues/2603, because sometimes people reported they were running the application multiple times (by mistake) and this resulted in inability to login and even DB corruption (?, not confirmed).
Also, I discussed this effort in one of the recent meetings and nobody provided any objections.

**Comment by pinebit** [comment]
> @pinebit would u mind to put a description for the functionality that is planned to be introduced in this PR? Having empty descriptions is confusing and not informative. Thanks

Added description. Let me know if the description is unclear.

**Comment by pinebit** [comment]
> Wouldnt this stop us from being able to login into more than one account?

is there a description anywhere how I can do this? thanks!

**Comment by pinebit** [comment]
I agree with @borismelnik that we should implement this for application, not for window. Will do this next and update the PR.

**Comment by richard-ramos** [comment]
> is there a description anywhere how I can do this? thanks!

- if the accounts are not pointing to the same network, (i.e. one on ropsten and the other on mainnet),  you can open two instances of desktop with each account. 
- If you are running desktop executable on separate folders you can start N instances using the same network. 
 
Having said that, I don't know if running more than one instance is something that we want. 
cc: @iurimatias and @anastasiyaig

**Comment by iurimatias** [comment]
> Having said that, I don't know if running more than one instance is something that we want.
> cc: @iurimatias and @anastasiyaig

Multiple instances is something we most definitely want to support, so long they are pointing to different data folders, myself I typically run a personal production version installed in the system, and the dev version, concurrently.

**Comment by anastasiyaig** [comment]
same here, i run a lot of instances, like PRs, master builds and release builds in parallel , moreover this ability to run them in different data folders is something that was designed and developed by @michaelsbradleyjr  specifically and per request

**Comment by pinebit** [comment]
To clarify: this implementation **does allow** you to run multiple instances given they are located in **different folders**.
However, if the user is running the app twice from the same application folder, then it will prevent second instance to start and activate the first one.
Sorry if I had not make it clear in the description.

**Review by iurimatias — CHANGES_REQUESTED** [review]
so this PR, I tested it, then merged the dotherside/nimqml PRs and updated here, tested one last time but unfortunatly ran into an issue, when I open the app the first time it works, but the second time it does not:

```
STATUS_PORT=30311 make run V=1
INFO [08-05|16:27:13.997] Status backend initialized               backend=geth version=0.83.4 commit=c4a71f81
QLocalServer::listen( "/tmp/f568c6621495cbc340070a6894d623da" ) failed
INF 2021-08-05 16:27:14-04:00 Terminating the app as the second instance topics="main" tid=5794816 file=nim_status_client.nim:69
```

**Comment by pinebit** [comment]
@iurimatias made a fix, need to review https://github.com/status-im/dotherside/pull/35
also I rebased this on the latest master.

I tested this on Ubuntu 20. I found that application is crashing when closing the main window (Segmentation fault).
This resulted in the pipe file remaining in the file system. The fix I made is about removing that pipe file if we were unable to connect (the second instance).
Another issue I found was that the app window cannot restore/show if it was minimized.
Looks like on Linux we have several other issues...
To sum up:
1. segfault on application close
2. app window does not show/raise programmatically if it was minimized by user

Other than that, SingleInstance works on Ubuntu, even after application was crashed.
Please verify at your end. Thanks!

**Comment by pinebit** [comment]
Regarding the crash on Linux, stack trace is not really useful...
```
#0  0x0000000000000000 in  ()
#1  0x00007ffff51ce07c in QMetaObject::invokeMethod(QObject*, char const*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) () at /home/pinebit/Qt/5.14.2/gcc_64/lib/libQt5Core.so.5
#2  0x0000555555e30a9b in dos_signal ()
#3  0x00005555556fcd78 in signal_handler__MS44G2MTvbYc9cYl9bV7Umww (receiver=0x555556279020, signal_0=0x7fffb0002640 "{\"type\":\"stats\",\"event\":{\"uploadRate\":0,\"downloadRate\":0}}", slot=0x555555f685ad "receiveSignal")
    at /home/pinebit/status-desktop/vendor/nimqml/src/nimqml.nim:41
#4  0x0000555555e1e9e8 in colonanonymous___qFVXMmdH9c9awMKwptoPjFrA (p0=0x7fffb0002640 "{\"type\":\"stats\",\"event\":{\"uploadRate\":0,\"downloadRate\":0}}") at /home/pinebit/status-desktop/src/nim_status_client.nim:237
#5  0x00007ffff3bd5ee9 in StatusServiceSignalEvent (jsonEvent=<optimized out>) at signals.c:15
#6  0x00007ffff3bd5eab in _cgo_d42b8c8f70e3_Cfunc_StatusServiceSignalEvent (v=0xc000b3fe80) at cgo-gcc-prolog:65
#7  0x00007ffff2accfc8 in runtime.asmcgocall () at /usr/local/go/src/runtime/asm_amd64.s:656
#8  0x00007ffff2a9d30b in runtime.park_m (gp=0x13b13432) at /usr/local/go/src/runtime/proc.go:2851
#9  0x000000c000caf080 in  ()
#10 0x0000000013b13432 in  ()
#11 0x00007fffb0000020 in  ()
#12 0x0000000000800000 in  ()
#13 0x0000000000000000 in  ()
```

**Comment by iurimatias** [comment]
tested this again and this is the issue I found, steps are:

1) run the 'system' app i.e the one installed in `/Applications` for personal use etc..
2) then run this branch `STATUS_PORT=30311 make run V=1`  (after compiling etc..)

expected result: dev app should still work side by side with system one, they don't share the same datadir after all
actual result:

```
QLocalServer::listen( "/tmp/f568c6621495cbc340070a6894d623da" ) failed
INF 2021-09-06 17:17:36-04:00 Terminating the app as the second instance topics="main" tid=15034740 file=nim_status_client.nim:69
```

**Comment by michaelsbradleyjr** [comment]
Heads up that `STATUS_PORT` no longer has any effect or use since mid February of this year:

https://github.com/status-im/status-desktop/pull/1858

A couple of months after that change, I added a comment in the Makefile to help clear up any confusion on that point:

https://github.com/status-im/status-desktop/blob/master/Makefile#L450-L451

**Comment by iurimatias** [comment]
I did this again, this time, after compiling doing `make pkg` and installing that app and still face a similar issue:

1. remove vendor, compile everything
2. `make pkg`

**Comment by iurimatias** [comment]
I did this again, this time, after compiling doing `make pkg` and installing that app and still face a similar issue:

1. remove vendor, compile everything
2. `make pkg`
3. open dmg and install app
4. start app
5. go back to terminal and run dev app `make run V=1` or `./bin/nim_status_client`

result, app quits with this message:

```
INFO [09-09|16:17:51.087] Status backend initialized               backend=geth version=0.85.0 commit=192ca6e3
INF 2021-09-09 16:17:51-04:00 Terminating the app as the second instance topics="main" tid=17684484 file=nim_status_client.nim:70
```

**Comment by pinebit** [comment]
> I did this again, this time, after compiling doing `make pkg` and installing that app and still face a similar issue:
> 
> 1. remove vendor, compile everything
> 2. `make pkg`
> 3. open dmg and install app
> 4. start app
> 5. go back to terminal and run dev app `make run V=1` or `./bin/nim_status_client`
> 
> result, app quits with this message:
> 
> ```
> INFO [09-09|16:17:51.087] Status backend initialized               backend=geth version=0.85.0 commit=192ca6e3
> INF 2021-09-09 16:17:51-04:00 Terminating the app as the second instance topics="main" tid=17684484 file=nim_status_client.nim:70
> ```

Thank you for testing! Did you have any other status instances/processes running? Or this is just for the first instance?
Unfortunately I don't have a Mac to try reproducing this, but I will try on Linux, I hope they should behave identically.

**Comment by iurimatias** [comment]
it worked!

**Comment by anastasiyaig** [comment]
it works to me as well on mac
